googletest/feature_support_test.cpp

```diff
@@ -0,0 +1,1313 @@
+
+    INIT_FEATURES();
+    EXPECT_FALSE(features.TileBasedRenderer(0));
+    EXPECT_TRUE(features.UMA(0));
+    EXPECT_FALSE(features.CacheCoherentUMA(0));
+    EXPECT_FALSE(features.IsolatedMMU(0)); // If Architecture1 is not available, IsolatedMMU should not be available
+}
+
+// Test on devices with more than one graphics node
+TEST_F(FeatureSupportTest, ArchitectureMulticore)
+{
+    device->SetNodeCount(3);
+    device->m_Architecture1Available = true;
+    device->m_TileBasedRenderer = {false, true, true};
+    device->m_UMA = {true, false, false};
+    device->m_CacheCoherentUMA = {false, false, true};
+    device->m_IsolatedMMU = {false, true, false};
+
+    INIT_FEATURES();
+    
+    EXPECT_FALSE(features.TileBasedRenderer(0));
+    EXPECT_TRUE(features.TileBasedRenderer(1));
+    EXPECT_TRUE(features.TileBasedRenderer(2));
+    
+    EXPECT_TRUE(features.UMA(0));
+    EXPECT_FALSE(features.UMA(1));
+    EXPECT_FALSE(features.UMA(2));
+
+    EXPECT_FALSE(features.CacheCoherentUMA(0));
+    EXPECT_FALSE(features.CacheCoherentUMA(1));
+    EXPECT_TRUE(features.CacheCoherentUMA(2));
+
+    EXPECT_FALSE(features.IsolatedMMU(0));
+    EXPECT_TRUE(features.IsolatedMMU(1));
+    EXPECT_FALSE(features.IsolatedMMU(2));
+}
+
+// 2: Feature Levels
+// Basic test with a high feature level
+TEST_F(FeatureSupportTest, FeatureLevelBasic)
+{
+    device->m_FeatureLevel = D3D_FEATURE_LEVEL_12_2;
+    
+    INIT_FEATURES();
+    EXPECT_EQ(features.HighestFeatureLevel(), D3D_FEATURE_LEVEL_12_2);
+}
+
+// Test through all supported feature levels
+TEST_F(FeatureSupportTest, FeatureLevelAll)
+{
+    std::vector<D3D_FEATURE_LEVEL> allLevels = 
+    {
+        D3D_FEATURE_LEVEL_1_0_CORE,
+        D3D_FEATURE_LEVEL_9_1,
+        D3D_FEATURE_LEVEL_9_2,
+        D3D_FEATURE_LEVEL_9_3,
+        D3D_FEATURE_LEVEL_10_0,
+        D3D_FEATURE_LEVEL_10_1,
+        D3D_FEATURE_LEVEL_11_0,
+        D3D_FEATURE_LEVEL_11_1,
+        D3D_FEATURE_LEVEL_12_0,
+        D3D_FEATURE_LEVEL_12_1,
+        D3D_FEATURE_LEVEL_12_2
+    };
+
+    for (unsigned int i = 0; i < allLevels.size(); i++) 
+    {
+        device->m_FeatureLevel = allLevels[i];
+        INIT_FEATURES();
+        EXPECT_EQ(features.HighestFeatureLevel(), allLevels[i]);
+    }
+}
+
+// Test with a feature level higher than the current highest level
+TEST_F(FeatureSupportTest, FeatureLevelHigher)
+{
+    device->m_FeatureLevel = (D3D_FEATURE_LEVEL)(D3D_FEATURE_LEVEL_12_2 + 1);
+    INIT_FEATURES();
+    EXPECT_EQ(features.HighestFeatureLevel(), D3D_FEATURE_LEVEL_12_2);
+}
+
+// 3: Format Support
+// Forward call to the old API. Basic correctness check
+TEST_F(FeatureSupportTest, FormatSupportPositive)
+{
+    device->m_FormatSupport1 = D3D12_FORMAT_SUPPORT1_BUFFER | D3D12_FORMAT_SUPPORT1_TEXTURE3D | D3D12_FORMAT_SUPPORT1_SHADER_SAMPLE_COMPARISON | D3D12_FORMAT_SUPPORT1_DECODER_OUTPUT;
+    device->m_FormatSupport2 = D3D12_FORMAT_SUPPORT2_UAV_ATOMIC_ADD | D3D12_FORMAT_SUPPORT2_TILED;
```

Comment:
The return value for format support (as well as other functions that involves COMMAND_ID or DXGI_FORMAT) is arbitrary. The test suite only makes sure that input and output values are forwarded correctly. In the manual test suite I will use meaningful data for testing.

Reply:
This would be a good note to capture in a comment.

General principle: any time you find yourself writing comments on a pull request to explain the code, and that comment refers to how the code in question will always be rather than just some temporary state that will change soon, consider adding a source comment instead so that the PR comment is not needed.

Reply:
I've added this as an comment in the new commit.